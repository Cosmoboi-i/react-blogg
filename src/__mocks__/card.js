export const mockCard = {
  id: 1,
  date: "2nd Januray, 2018",
  reading_time: "2 mins",
  title: "The future of abstract art and the culture ...",
  description:
    "Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...",
  claps: 10,
  liked: false,
  image: "abstract.png",
};

export const mockKey = "6955f78e-888d-4109-a2ff-abd88cf3c08f";
