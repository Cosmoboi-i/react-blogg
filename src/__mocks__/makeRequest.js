export const mockReq = {
  body: {
    claps: 10,
    liked: false,
  },
};

export const mockRes = {
  data: {
    claps: 10,
    liked: false,
  },
};
