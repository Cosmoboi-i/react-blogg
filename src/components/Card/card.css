.card {
  height: 500px;
  width: 350px;
  border: 1px solid rgb(198, 198, 198);
  margin: 1.2em 0;
}

.image {
  height: 280px;
  width: 350px;
  border: 0 none;
}

.card-content {
  display: flex;
  flex-flow: column wrap;

  padding: 0.4em 0.6em;
  font-weight:lighter;
}

.date-time {
  display: flex;
  justify-content: space-between;

  color: rgb(198, 198, 198);
  font-size: smaller;
}

.card-text {
  padding: 0 0.4em 0 0;
  
}

.card-heading {
  font-size: x-large;
  margin: 0.6em 0;
  height: 2.4em;
  overflow: hidden;              
  text-overflow: ellipsis;
}

.card-desc {
  font-size: normal;
  color: rgb(156, 156, 156);
  height: 3.5em;
  width: 320px;
  overflow: hidden;              
  text-overflow:    ellipsis;
}

.hr {
  width: 100%;
  margin: 0.6em 0;
  border: 0.5px solid rgb(198, 198, 198);
}

.clap {
  display: flex;
  align-items: center;
}

.card-reaction {
  display: flex;
  justify-content: space-between;
}

.card-reaction img {
  height:1.5em;
  margin-right: 0.1em;
}

.clap span {
  align-self: center;
  color: rgb(157, 157, 157);
}