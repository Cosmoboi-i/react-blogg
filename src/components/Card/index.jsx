import React, { useEffect, useState } from "react";
import "./card.css";
import PropTypes from "prop-types";
import makeRequest from "../../utils/makeRequest";
import { UPDATE_BLOG_DATA } from "../../constants/apiEndPoints";
import { MONTHS } from "../../constants/months";

const svgPath = "./assets/Icons/";

export default function Card(props) {
  const [liked, setLiked] = useState(props.liked);
  const [claps, setClaps] = useState(props.claps);

  const dateObj = new Date(props.date);
  const date =
    dateObj.getDate() +
    " " +
    MONTHS[dateObj.getMonth()] +
    "," +
    " " +
    dateObj.getFullYear();

  //console.log(date);

  useEffect(() => {
    (async function fetchData() {
      const data = { claps, liked };
      await makeRequest(UPDATE_BLOG_DATA(props.id), { data });
      //console.log(res);
    })();
  });

  return (
    <div className="card" data>
      <img src={props.image} className="image" />
      <div className="card-content">
        <div className="date-time">
          <div>{date}</div>
          <div>{props.reading_time}</div>
        </div>
        <div className="card-text">
          <div className="card-heading">{props.title}</div>
          <div className="card-desc">{props.description}</div>
        </div>
        <hr className="hr" />
        <div className="card-reaction">
          <div className="clap" data-testid="clap">
            <img
              src={svgPath + "clapping.svg"}
              alt="clap"
              onClick={() => setClaps(claps + 1)}
            />
            <span data-testid="clap-count">{claps}</span>
          </div>
          <img
            src={svgPath + (liked ? "heart-red.svg" : "heart-black.svg")}
            alt="heart"
            className="card-heart"
            onClick={() => setLiked(!liked)}
          />
        </div>
      </div>
    </div>
  );
}

Card.propTypes = {
  id: PropTypes.number.isRequired,
  image: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  reading_time: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  claps: PropTypes.number.isRequired,
  liked: PropTypes.bool.isRequired,
};
