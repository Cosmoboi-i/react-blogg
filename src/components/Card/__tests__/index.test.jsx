import React from "react";
import {
  render,
  fireEvent,
  getByAltText,
  within,
} from "@testing-library/react";
import Card from "..";
import { mockCard, mockKey } from "../../../__mocks__/card";

describe("Card", () => {
  it("should render the Card", () => {
    const { asFragment } = render(<Card key={mockKey} {...mockCard} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should set heart icon to red when clicked and it is black", () => {
    const screen = render(<Card key={mockKey} {...mockCard} />);
    const heart = screen.getByAltText("heart");

    expect(heart.src).toContain("heart-black.svg");
    fireEvent.click(heart);
    expect(heart.src).toContain("heart-red.svg");
  });

  it("should set heart icon to black when clicked and it is red", () => {
    const screen = render(<Card key={mockKey} {...mockCard} />);
    const heart = screen.getByAltText("heart");

    fireEvent.click(heart);
    expect(heart.src).toContain("heart-red.svg");
    fireEvent.click(heart);
    expect(heart.src).toContain("heart-black.svg");
  });

  it("should increase clap count by 1 when clicked on clap icon", () => {
    const screen = render(<Card key={mockKey} {...mockCard} />);
    const clap = screen.getByAltText("clap");
    const clapCount = screen.getByTestId("clap-count");
    expect(clapCount.textContent).toBe("10");
    fireEvent.click(clap);
    expect(clapCount.textContent).toBe("11");
  });
});
