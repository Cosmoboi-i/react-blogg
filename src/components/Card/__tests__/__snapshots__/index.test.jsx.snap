// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render the Card 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <img
      class="image"
      src="abstract.png"
    />
    <div
      class="card-content"
    >
      <div
        class="date-time"
      >
        <div>
          NaN undefined, NaN
        </div>
        <div>
          2 mins
        </div>
      </div>
      <div
        class="card-text"
      >
        <div
          class="card-heading"
        >
          The future of abstract art and the culture ...
        </div>
        <div
          class="card-desc"
        >
          Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...
        </div>
      </div>
      <hr
        class="hr"
      />
      <div
        class="card-reaction"
      >
        <div
          class="clap"
          data-testid="clap"
        >
          <img
            alt="clap"
            src="./assets/Icons/clapping.svg"
          />
          <span
            data-testid="clap-count"
          >
            10
          </span>
        </div>
        <img
          alt="heart"
          class="card-heart"
          src="./assets/Icons/heart-black.svg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
